core/src/utils/prove_distributed/checkpoints.rs

```diff
@@ -0,0 +1,154 @@
+use std::sync::{mpsc::SyncSender, Arc};
+
+pub use crate::{air::PublicValues, runtime::Program, stark::RiscvAir};
+
+use crate::{
+    runtime::{ExecutionRecord, NoOpSubproofVerifier, Runtime},
+    stark::{MachineProver, MachineRecord},
+    utils::{baby_bear_poseidon2::Val, BabyBearPoseidon2, SP1CoreOpts},
+};
+
+use super::Checkpoint;
+
+fn trace_checkpoint(
+    program: Program,
+    checkpoint: Checkpoint,
+    opts: SP1CoreOpts,
+) -> (Vec<ExecutionRecord>, Checkpoint) {
+    let mut runtime = Runtime::recover(program, checkpoint, opts);
+
+    runtime.subproof_verifier = Arc::new(NoOpSubproofVerifier);
+
+    let (events, _) =
+        tracing::debug_span!("runtime.trace").in_scope(|| runtime.execute_record().unwrap());
+
+    let state = runtime.state.clone();
+
+    (events, state)
+}
+
+pub fn process<P: MachineProver<BabyBearPoseidon2, RiscvAir<Val>>>(
+    prover: &P,
+    program: &Program,
+    checkpoint: Checkpoint,
+    nb_checkpoints: usize,
+    state: PublicValues<u32, u32>,
+    opts: SP1CoreOpts,
+    records_tx: SyncSender<Vec<ExecutionRecord>>,
+    deferred: &mut ExecutionRecord,
+    is_deferred: bool,
+) {
+    if is_deferred {
+        process_deferred(program, checkpoint, state, opts, records_tx, deferred);
+    } else {
+        process_regular(
+            prover,
+            program,
+            checkpoint,
+            nb_checkpoints,
+            state,
+            opts,
+            records_tx,
+            deferred,
+        );
+    }
+}
+
+fn process_regular<P: MachineProver<BabyBearPoseidon2, RiscvAir<Val>>>(
+    prover: &P,
+    program: &Program,
+    mut checkpoint: Checkpoint,
+    nb_checkpoints: usize,
+    mut state: PublicValues<u32, u32>,
+    opts: SP1CoreOpts,
+    records_tx: SyncSender<Vec<ExecutionRecord>>,
+    deferred: &mut ExecutionRecord,
+) {
```

Comment:
This probably needs a comment or a link to a markdown document that explains the shared mutable state.

From the code, you have shards that are logical and mapped to execution_shards.
It seems like the distribution happens ahead of time but then after records are processed this execution shard value is incremented.

Given the potential complexity, I would cleanly have only execution functions and only orchestrating functions. And execution functions should only return a status code. I fear mixing processing and state updates will lead to maintenance burden down the line.

This can be refactored for a follow-up PR.

Reply:
I agree about needing some more documentation.

This code is an adaptation of the classical sp1's proving process taken from https://github.com/taikoxyz/sp1/blob/main/core/src/utils/prove.rs#L234

While I changed the code organization a bit to better fit our needs, I think we shouldn't tamper too much with the substance of the original code or face a possible complete re implementation of the solution whenever they make some breaking changes upstream. 

This whole `deferred` mechanism is a new system that avoids putting too many heavy public values on each shard. It does improve the performances a bit, but implies a new layer of synchronization and value-sharing among the workers and the orchestrator. If we decide to go down the road of implementing a custom solution we could get rid of that mechanism and simplify the execution path drastically. But once again this has a greater maintenance cost.